### PR TITLE
libbpfgo: Add map batch operations

### PR DESCRIPTION
**Pull request by derekparker**
This patch adds a few batch operations for maps including:

Lookup, LookupAndDelete, Update and Delete.

### CODE REVIEW

File: Makefile
Comment (rafaeldtinoco) [inline Makefile:80]: could you, pls, keep the indentation here as it was ? I'm going to "fix" the Makefile style soon, not only for this line but for other lines as well.

Comment (derekparker) [reply Makefile:80]: Ack, no problem!

**Review by rafaeldtinoco — DISMISSED** [review]
LGTM. We should open an issue to transform the "selftest/build" into a proper selftest/XXX directory. @grantseltzer WDYT ?

Besides, I think we should "unify" the Makefiles into a single Makefile at the root dir (just like we recently did to tracee). 

Nevertheless, I think we should accept this contribution, which was great.

@grantseltzer I'll let you squash & merge this if you agree.

File: libbpfgo.go
Comment (grantseltzer) [inline libbpfgo.go:619]: errno needs to be checked here, you can use syscall.Errno like elsewhere in libbpfgo.

File: libbpfgo.go
Comment (grantseltzer) [inline libbpfgo.go:643]: errno needs to be checked here, you can use syscall.Errno like elsewhere in libbpfgo.

File: libbpfgo.go
Comment (grantseltzer) [inline libbpfgo.go:687]: errno needs to be checked here, you can use syscall.Errno like elsewhere in libbpfgo.

File: libbpfgo.go
Comment (grantseltzer) [inline libbpfgo.go:672]: errno needs to be checked here, you can use syscall.Errno like elsewhere in libbpfgo.

File: Makefile
Comment (grantseltzer) [inline Makefile:84]: can you add `-v` in here while you're at it?

**Review by grantseltzer — CHANGES_REQUESTED** [review]
Thanks so much for this contribution Derek! 

In addition to my inline comments, can I ask that you move your tests into `selftests`? It shouldn't make much of a change, just copy the directory structure of one of the other selftests.

I'm not entirely sure why we still have the test that's in `libbpfgo_test.go` (Test_LoadAndAttach), that file should contain tests that don't shell out, things like helper functions and whatnot. I'm going to create an issue to track removing/replacing that one.

File: libbpfgo.go
Comment (derekparker) [reply libbpfgo.go:602]: That SGTM, will update.

Comment (derekparker) [reply libbpfgo.go:602]: Actually, the way I would implement this would be to return `map[[]byte][]byte` because we don't know what type the keys will be (or size). The problem with this is a `[]byte` is not comparable so cannot be used as a key. I suppose we could use a `map[interface{}][]byte` and that should work (haven't tested it) but feels kind of weird to me, and adds a type assertion to the callers code.

I'm happy to implement the latter if that's what you'd prefer, however this interface matches what happens in libbpf (although admittedly it's not very Go-like).

As shown in the test it's fairly straightforward still to iterate through keys / values using a single iterator var to keep track of where you are.

File: libbpfgo.go
Comment (derekparker) [reply libbpfgo.go:619]: Apologies, I followed the example of the other BPFMap methods (such as `Pin`, `UnPin`, etc...) which don't do any extra checking on the returned error than what I've done. Do you have another specific example or pattern you'd like me to follow or extra checks to perform?

File: Makefile
Comment (derekparker) [reply Makefile]: Done.

**Comment by derekparker** [comment]
Hey @grantseltzer thanks for the feedback!

I've addressed most of your review except for that which I've commented on. Also I've moved the test over to it's own directory in `selftest/map-batch` and removed the corresponding tests in `libbpfgo_test.go`.

File: libbpfgo.go
Comment (grantseltzer) [reply libbpfgo.go:602]: Good points, best to leave it as is.

File: libbpfgo.go
Comment (grantseltzer) [reply libbpfgo.go:619]: I had an answer ready to this but then after carefully examining libbpf code I got confused. I asked on the mailing list how functions which use `libbpf_err` vs `libbpf_err_errno` are supposed to be checked, let's wait for a response from Andrii before proceededing:

https://lore.kernel.org/bpf/CAO658oUqd5=B3zkDhm2jVQxG+vEf=2CE7WimXHqgcH+m0P=k_Q@mail.gmail.com/T/#u

I'm sure there's functions in libbpfgo that aren't correct.

Comment (grantseltzer) [reply libbpfgo.go:619]: @yanivagman You probably have a good idea

File: libbpfgo.go
Comment (derekparker) [reply libbpfgo.go:602]: Sounds good!

File: libbpfgo.go
Comment (derekparker) [reply libbpfgo.go:619]: Ack, sounds good to me!

Comment (derekparker) [reply libbpfgo.go:619]: If I understand correctly it looks like Andrii responded saying that the returned value can be used directly without checking errno unless it's an API which returns a pointer, in which case errno must be checked.

Are you reading the response the same?

Comment (grantseltzer) [reply libbpfgo.go:619]: Yes I read it the same, which means the way you're doing it is correct. Though perhaps we can extract the error code to determine what the issue was. I think that's what using the `syscall.Errno` type does, but I will investigate in the morning.

File: selftest/map-batch/main.go
Comment (grantseltzer) [inline selftest/map-batch/main.go:82]: Can you explain more of what's going on here? Why is there two steps? I see the documentation for BPF_MAP_LOOKUP_BATCH in uapi/linux/bpf.h but I don't fully understand why it's designed this way.

File: selftest/map-batch/main.bpf.c
Comment (derekparker) [reply selftest/map-batch/main.bpf.c]: Removed.

File: selftest/map-batch/main.bpf.c
Comment (derekparker) [reply selftest/map-batch/main.bpf.c]: Done, forgot to do that when swapping over.

File: selftest/map-batch/main.bpf.c
Comment (derekparker) [reply selftest/map-batch/main.bpf.c]: Done.

File: selftest/map-batch/main.go
Comment (derekparker) [reply selftest/map-batch/main.go:65]: Ack, that's my bad forgot to update when swapping over from the other test.

File: selftest/map-batch/main.go
Comment (derekparker) [reply selftest/map-batch/main.go]: Removed.

File: selftest/map-batch/main.go
Comment (derekparker) [reply selftest/map-batch/main.go]: Removed.

File: selftest/map-batch/main.go
Comment (derekparker) [reply selftest/map-batch/main.go:60]: Yeah, so the API is such that you can iterate over all the keys/values in batches. Essentially you provide a pointer to the last key seen, the next key to start with, and a buffer for keys/values.

That way if you have a map with max size of 1000 for example you can iterate over the entire thing in chunks instead of loading everything into memory. It's kind of a confusing API to be honest it took me a bit to wrap my head around it when implementing.

Comment (grantseltzer) [reply selftest/map-batch/main.go:60]: Got it - that took some time to understand. Definitely worth documenting upstream in libbpf.

Comment (grantseltzer) [reply selftest/map-batch/main.go:60]: After that this is good to go :+1:

Comment (derekparker) [reply selftest/map-batch/main.go:60]: > Got it - that took some time to understand. Definitely worth documenting upstream in libbpf.

Yes! I literally had to read through the test code to understand how to use the API.

Comment (derekparker) [reply selftest/map-batch/main.go:60]: Ok, let me know if the documentation I added is clear enough.

File: libbpfgo.go
Comment (grantseltzer) [inline libbpfgo.go:606]:
```suggestion
// and repeat until nextKey is nil.
```

**Comment by grantseltzer** [comment]
Made a small edit I didn't want to bother you with. This looks good to me! Thanks so much @derekparker !

**Review by grantseltzer — APPROVED** [review]

**Pull request merged**
